fdbserver/include/fdbserver/MovingWindow.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * MovingWindow.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#ifndef FOUNDATIONDB_MOVINGWINDOW_H
+#define FOUNDATIONDB_MOVINGWINDOW_H
+
+#include <limits.h>
+#include "flow/Deque.h"
+#include "fdbserver/Knobs.h"
+
+// Perfomed as the rolling window to calculate average change rates in the past <interval>
+// e.g., we may use it in "MovingData" Trace to show average moving bytes rate by DD.
+template <class T>
+class MovingWindow {
+private:
+	T previous;
+	T total;
+	// To avoid having a super large Deque which may lead OOM, we set a maxSize for it.
+	// Actually, Deque has its own Deque::max_size = 1 << 30, We may narrow it down here.
+	int maxDequeSize;
+	Deque<std::pair<double, T>> updates; // pair{time, numeric}
+	double interval;
+	// Updated when initialization Or pop() due to full Deque
+	double previousPopTime;
+
+	void pop() {
+		previous += updates.front().second;
+		updates.pop_front();
+	}
+
+public:
+	MovingWindow() = default;
+	explicit MovingWindow(double timeWindow)
+	  : previous(0), total(0), maxDequeSize(SERVER_KNOBS->MOVING_WINDOW_SAMPLE_SIZE / sizeof(std::pair<double, T>)),
+	    interval(timeWindow), previousPopTime(now()) {}
+
+	T getTotal() const { return total; }
+
+	double getAverage() {
```

Comment:
This is similar to Smoother. It's clearly a bit different, but we might want to articulate why we didn't use that

Reply:
`MovingWindow` would just add up all the samples, while `Smoother` would modify samples to make it continuous.
In some cases(e.g. `MovingData` Trace log), we want to show the accurate dd movement rate in past 120 seconds.

Reply:
I wonder if you can comment more on why we want the properties of the moving window's average rather than the smoothed rate? Do we have some usage of this data that is motivating that desire? 

Based on that, it would probably be useful to give a clear motivation for choosing between the two as documentation in the code. We had a similar concern with the various methods we implemented to collect samples/histograms where it wasn't obvious what the benefit of one over the other was.

Reply:
I've created a new PR(https://github.com/apple/foundationdb/pull/9812) which comments on the usage and difference of `MovingWindow` and `Smoother`. Does it make sense to merge this PR at first and then cherry-pick the other one later?

Reply:
Sounds good to me